### DISCUSSION TITLE
ASP.NET Classic Web Projects and app/web.config

### DISCUSSION THREAD

**Discussion by tibitoth** [microsoft/aspire#4280] [Q&A] [answered]
Hi Team, 
Aspire is a great product, congratulations!

One question and feel free to move this to a discussion if it's not an issue.

Currently I have to work on an old project with a bunch of ASP.NET Classic projects (.NET FW, MVC, WCF, app.config. web.config etc) and the onboarding process is a disaster. Could I use .NET Aspire to make this as streamlined as in your demos?

(I know out customer should migrate to ASP.NET Core but that's a different story)

**Comment by davidfowl** [comment]
The orchestrator project can launch classic .NET Framework projects while running in visual studio but there are lots of gaps in the experience (ones that do not plan to fill):
- It's unclear what the open telemetry support looks like for .NET Framework applications
- We pass connection strings using the IConfiguration based patterns. Environment variables don't work well with class .NET Framework's configuration system.
- Aspire components target .NET 8 (these are the packages that run in your application)
- Deployment won't work. It's not even clear what it would do.

So you can launch and debug your application and dependencies, everything else you'll have to build. At least its not yaml so you can! 😄

**Reply by tdhatcher** [reply]
Seeing what options are currently possible and how to go about this would be great. 

The local hosting aspect is probably enough for most folks' use cases. I suspect most in that camp probably wouldn't be changing their existing deployment or looking for service discoverability or assistance of Aspire components because the application is in a maintenace/support mode (no investment).

It's probable that this appeals to devs where its not practical or feasible within their organization to dedicate additional resources simply to upgrade a legacy application that is likely to be tombstoned in a couple years or not impact the user experience or economics. However some poor dev team (with devs joining and leaving) still must handle the hot potato legacy application as priorities change in an organization. So it has value in reducing onboarding complexity.

I think being able to have the legacy app plug into the orchestrator at a basic level could help derisk the upgrade path and make the case for its feasibility and investment since you could reduce the churn as much as possible running the entire ecosystem locally. It opens the more to implement a fig strangler or big bang rewrite run both versions concurrently and execute tests to confirm behavior parity.

**Reply by Sebbs128** [reply]
tdhatcher's last point is where I think the orchestration and service discovery with legacy ASP.NET would be very beneficial: migrating legacy ASP.NET apps to ASP.NET Core, especially when following the incremental migration approach (with YARP).

In a simple case of migrating a web app + REST API and a database, there's 4 projects that need to be launched (aspnetcore web app + aspnet web app, aspnetcore api + aspnet api), and at least two services to be shared via service discovery/config (database connection string, base URI for the aspnetcore api).

If the orchestrator can already launch ASP.NET projects, then that's a great first step (it's not obvious from docs or aspire-samples how to do this, but I'm guessing as an `ExecutableResource`?). Without `IConfiguration` I guess writing into the web.config `connectionStrings` and `appSettings` would be the way to go (where does xml rank vs yaml on the markup language tier list? 😛). If `IConfiguration` is being used in the ASP.NET apps (eg. via a service locator), then the injection via environment variables should work.

Unfortunately I imagine this would all need to be a community supported+packaged effort though.

**Reply by davidfowl** [reply]
Sufficiently motivated individuals could build this with the building blocks implemented today. This isn't a scenario we are prioritizing as part of aspire, but as an example of how you can start to out something like this, here's an IIS express based resource https://github.com/dotnet/aspire/discussions/4383#discussioncomment-9674132 that launches an executable.

In the [Aspire Discord](https://aka.ms/aspire/discord) there are a couple of customers that are building some crazy extensions to work with legacy applications (writing files to disk and launching etc). Finding a group of likeminded individuals to build something is a great way to get this kick started 😄.

Also https://github.com/dotnet/aspire/discussions/1439#discussioncomment-9178619

**Comment by Sebbs128** [comment, marked as answer]
@tibitoth @tdhatcher I've been working on a proof of concept to see how some of this will play out. It's currently published to https://github.com/Sebbs128/AspireNetFramework. It's based a little on the IIS Express-based resource David Fowler shared earlier (but pulls the existing site configs VS uses from `.vs/{solution name}/config/applicationhost.config`.

I'm definitely interested in trying to develop this further in a community-maintained library, if either of yourselves are also interested (or know of people who are).

**Reply by davidfowl** [reply]
@Sebbs128 we just introduced the concept of a publisher (still in development) but it would be interesting to flesh out an iis based publisher/deployer.  

Here's the docker compose one:

https://github.com/dotnet/aspire/tree/b26d18a7f92d220203b68a18fd30626e72e3b704/src/Aspire.Hosting.Docker

**Reply by CZEMacLeod** [reply]
I have full* support for ASP.NET Framework using SDK projects, including debugging and cutover using SystemWeb-Adapters, and EF6 connections at https://github.com/CZEMacLeod/C3D.Extensions.Aspire.IISExpress
I haven't done anything with publishing, as we have existing infra for this for the legacy project, but would be happy to take contributions for integrating that.
*There are some limitations discussed in https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb

**Reply by davidfowl** [reply]
> I haven't done anything with publishing, as we have existing infra for this for the legacy project, but would be happy to take contributions for integrating that.

Well it hasn't shipped yet 😄

**Reply by davidfowl** [reply]
It shipped but we are still evolving the model. Still, I think it would be good to experiment if there's still interest.
